File: androidApp/src/main/java/com/dj/kmm/android/presentation/recipe_list/RecipeListViewModel.kt
package com.dj.kmm.android.presentation.recipe_list

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dj.kmm.domain.model.GenericMessageInfo
import com.dj.kmm.domain.model.NegativeAction
import com.dj.kmm.domain.model.PositiveAction
import com.dj.kmm.domain.model.Recipe
import com.dj.kmm.domain.model.UIComponentType
import com.dj.kmm.domain.util.GenericMessageInfoQueueUtil
import com.dj.kmm.domain.util.Queue
import com.dj.kmm.interactors.recipe_list.SearchRecipes
import com.dj.kmm.presentation.recipe_list.FoodCategory
import com.dj.kmm.presentation.recipe_list.RecipeListEvents
import com.dj.kmm.presentation.recipe_list.RecipeListState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class RecipeListViewModel
@Inject
constructor(
    private val savedStateHandle: SavedStateHandle,
    private val searchRecipes: SearchRecipes,
) : ViewModel() {
    val state: MutableState<RecipeListState> = mutableStateOf(RecipeListState())

    init {
        onTriggerEvent(RecipeListEvents.LoadRecipes)

        // Test
        // val messageInfoBuilder = GenericMessageInfo.Builder()
        //     .id(UUID.randomUUID().toString())
        //     .title("Weird")
        //     .uiComponentType(UIComponentType.Dialog)
        //     .description("Test")
        //     .positive(positiveAction = PositiveAction(
        //         positiveBtnTxt = "Ok", onPositiveAction = {
        //             state.value = state.value.copy(query = "Kale")
        //             onTriggerEvent(RecipeListEvents.NewSearch)
        //         }
        //     ))
        //     .negative(NegativeAction("Cancel", onNegativeAction = {
        //         state.value = state.value.copy(query = "Chicken")
        //         onTriggerEvent(RecipeListEvents.NewSearch)
        //     }))
        // appendToMessageQueue(messageInfo = messageInfoBuilder)
    }

    fun onTriggerEvent(event: RecipeListEvents) {
        when (event) {
            is RecipeListEvents.LoadRecipes -> {
                loadRecipes()
            }
            is RecipeListEvents.NextPage -> {
                nextPage()
            }
            is RecipeListEvents.NewSearch -> {
                newSearch()
            }
            is RecipeListEvents.OnUpdateQuery -> {
                state.value = state.value.copy(query = event.query, selectedCategory = null)
            }
            is RecipeListEvents.OnSelectCategory -> {
                onSelectCategory(event.category)
            }
            is RecipeListEvents.OnRemovedHeadMessageFromQueue -> {
                removeHeadMessage()
            }
            else -> {
                appendToMessageQueue(
                    GenericMessageInfo.Builder()
                        .id(UUID.randomUUID().toString())
                        .title("Error")
                        .uiComponentType(UIComponentType.Dialog)
                        .description("Invalid Event")
                )
            }
        }
    }

    private fun loadRecipes() {
        searchRecipes.execute(
            page = state.value.page,
            query = state.value.query,
        ).onEach { dataState ->
            state.value = state.value.copy(isLoading = dataState.isLoading)
            dataState.data?.let { recipes ->
                appendRecipes(recipes)
            }
            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }
        }.launchIn(viewModelScope)
    }

    private fun onSelectCategory(foodCategory: FoodCategory) {
        state.value = state.value.copy(selectedCategory = foodCategory, query = foodCategory.value)
        newSearch()
    }

    private fun newSearch() {
        state.value = state.value.copy(page = 1, recipes = listOf())
        loadRecipes()
    }

    private fun nextPage() {
        state.value = state.value.copy(page = state.value.page + 1)
        loadRecipes()
    }

    private fun appendRecipes(recipes: List<Recipe>) {

        val curr = ArrayList(state.value.recipes)
        curr.addAll(recipes)
        state.value = state.value.copy(recipes = curr)
    }

    private fun appendToMessageQueue(messageInfo: GenericMessageInfo.Builder) {
        if (!GenericMessageInfoQueueUtil().doesMessageAlreadyExistInQueue(
                queue = state.value.queue, messageInfo = messageInfo.build()
            )
        ) {
            val queue = state.value.queue
            queue.add(messageInfo.build())
            state.value = state.value.copy(queue = queue)
        }
    }

    private fun removeHeadMessage() {
        try {
            val queue = state.value.queue
            queue.remove()
            state.value = state.value.copy(queue = Queue(mutableListOf())) // force recompose
        } catch (e: Exception) {
            //nothing to remove, queue is empty
        }
    }
}

File: iosApp/iosApp/presentation/recipe_list/RecipeListViewModel.swift
//
//  RecipeListViewModel.swift
//  iosApp
//
//  Created by dj on 2021/07/08.
//  Copyright © 2021 orgName. All rights reserved.
//

import SwiftUI
import shared

class RecipeListViewModel: ObservableObject {
    
    // dependencies
    let searchRecipes: SearchRecipes
    let foodCategoryUtil: FoodCategoryUtil
    
    // Observing
    @Published var state: RecipeListState = RecipeListState()
    
    init(searchRecipes: SearchRecipes, foodCategoryUtil: FoodCategoryUtil){
        self.searchRecipes = searchRecipes
        self.foodCategoryUtil = foodCategoryUtil
        
        // TODO("Perform Search")
        onTriggerEvent(stateEvent: RecipeListEvents.LoadRecipes())
    }
    
    func onTriggerEvent(stateEvent: RecipeListEvents){
        switch stateEvent {
        case is RecipeListEvents.LoadRecipes:
            loadRecipes()
        case is RecipeListEvents.NewSearch:
            newSearch()
        case is RecipeListEvents.NextPage:
            nextPage()
        case is RecipeListEvents.OnUpdateQuery:
            onUpdateQuery(query: (stateEvent as! RecipeListEvents.OnUpdateQuery).query)
        case is RecipeListEvents.OnSelectCategory:
            doNothing()
        case is RecipeListEvents.OnRemoveHeadMessageFromQueue:
            doNothing()
            
        default:
            doNothing()
        }
    }
    
    
    private func nextPage(){
        let currentState = self.state.copy() as! RecipeListState
        updateState(page: Int(currentState.page) + 1)
        loadRecipes()
    }
    
    private func loadRecipes(){
        let currentState = (self.state.copy() as! RecipeListState)
        do {
            try searchRecipes.execute(page: Int32(currentState.page), query: currentState.query)
                .collectCommon(
                    coroutineScope: nil,
                    callback: { dataState in
                        if dataState != nil {
                            let data = dataState?.data
                            let message = dataState?.message
                            let loading = dataState?.isLoading ?? false
                            self.updateState(isLoading: loading)
                            if data!=nil {
                                self.appendRecipes(recipes: data as! [Recipe])
                            }
                            if message != nil {
                                self.handleMessageByUIComponentType(message!.build())
                            }
                            
                        }
                    }
                )
        }catch{
            print("\(error)")
        }
    }
    
    private func newSearch(){
        resetSearchState()
        loadRecipes()
    }
    
    private func resetSearchState(){
        let currentState = (self.state.copy() as! RecipeListState)
        var foodCategory = currentState.selectedCategory
        if(foodCategory?.value != currentState.query){
            foodCategory = nil // unhighlight chip
        }
        self.state = self.state.doCopy(
            isLoading: currentState.isLoading,
            page: 1, // reset
            query: currentState.query,
            selectedCategory: foodCategory, // reset
            recipes: [], // reset
            bottomRecipe: currentState.bottomRecipe,
            queue: currentState.queue
        )
    }
    
    private func onUpdateQuery(query: String){
        updateState(query: query)
    }
    
    private func onUpdateBottomRecipe(recipe: Recipe){
        updateState(bottomRecipe: recipe)
    }
    
    func appendRecipes(recipes: [Recipe]){
        var currentState = (self.state.copy() as! RecipeListState)
        var currentRecipes = currentState.recipes
        currentRecipes.append(contentOf: recipes)
        
        self.state = self.state.doCopy(
            isLoading: currentState.isLoading,
            page: currentState.page,
            query: currentState.query,
            selectedCategory: currentState.selectedCategory,
            recipes: currentRecipes,
            bottomRecipe: currentState.bottomRecipe,
            queue: currentState.queue
        )
        currentState = self.state.copy() as! RecipeListState
        self.onUpdateBottomRecipe(recipe: currentState.recipes[currentState.recipes.count-1])
    }
    
    func shouldQueryNextPage(recipe: Recipe) -> Bool {
        let currentState = self.state.copy() as! RecipeListState
        if(recipe.id == currentState.bottomRecipe?.id){
            if(RecipeListState.Companion().RECIPE_PAGINATION_PAGE_SIZE * currentState.page <= currentState.recipes.count){
                if(!currentState.isLoading){
                    return true
                }
            }
        }
        return false
    }
    
    private func handleMessageByUIComponentType(_ message: GenericMessageInfo){
        // TODO("append to queue or None")
    }
    
    func doNothing(){
        //does nothing
    }
    
    func updateState(
        isLoading: Bool? = nil,
        page: Int? = nil,
        query: String? = nil,
        bottomRecipe: Recipe?=nil,
        queue: Queue<GenericMessageInfo>? = nil
    ){
        let currentState = (self.state.copy() as! RecipeListState)
        self.state = self.state.doCopy(isLoading: isLoading ?? currentState.isLoading,
                                       page: Int32(page ?? Int(currentState.page)),
                                       query: query ?? currentState.query,
                                       selectedCategory: currentState.selectedCategory,
                                       recipes: currentState.recipes,
                                       bottomRecipe: bottomRecipe ?? currentState.bottomRecipe,
                                       queue: queue ?? currentState.queue)
    }
}

